components/org.wso2.micro.integrator.initializer/src/main/java/org/wso2/micro/integrator/initializer/utils/ServiceCatalogUtils.java

```diff
@@ -424,6 +445,78 @@ private static boolean processMetadata(File tempDir, File metadataFolder, File m
+     * @throws NoSuchAlgorithmException could not find the MD% algorithm.
+     */
+    private static boolean processProxyServiceMetadata(File tempDir, File metadataYamlFolder,
+                                           Map<String, String> md5MapOfAllService) throws IOException
+            , ResolverException, NoSuchAlgorithmException {
+
+        String metaFileName = metadataYamlFolder.getName();
+        String proxyServiceName = metaFileName.substring(0,
+                metaFileName.indexOf(PROXY_SERVICE_SUFFIX + METADATA_FOLDER_STRING));
+        String proxyServiceVersion =
+                metaFileName.substring(metaFileName.lastIndexOf(METADATA_FOLDER_STRING) +
+                        METADATA_FOLDER_STRING.length());
+
+        // Create new folder in temp directory for this proxy service
+        File newMetaFile = new File(tempDir,  proxyServiceName + PROXY_SERVICE_SUFFIX
+                + "_v" + proxyServiceVersion);
+        File newYamlFile = new File(newMetaFile, METADATA_FILE_NAME);
+        File metadataYamlFile = new File(metadataYamlFolder, proxyServiceName + PROXY_SERVICE_SUFFIX
+                + METADATA_FILE_STRING + proxyServiceVersion + YAML_FILE_EXTENSION);
+
+        // Edit metadata yaml and add host details
+        String key = updateMetadataWithServiceUrl(metadataYamlFile);
+        String md5FromServer = md5MapOfAllService.get(key);
+
+        // Check WSDL file is fetched
+        if (!readProxyServiceWSDL(metadataYamlFile, metadataYamlFolder)) {
+            log.error("Could not find WSDL definition of proxy service: " + proxyServiceName);
+            return false;
+        }
+        File wsdlFile = new File(metadataYamlFolder, WSDL_FILE_NAME);
+
+        // Generate md5 values for verifier
+        String md5SumOfMetadata = getFileChecksum(metadataYamlFile);
+        String md5SumOfWSDL = getFileChecksum(wsdlFile);
+        String newMD5String = md5SumOfWSDL + md5SumOfMetadata;
+
+        // if proxy-service is not registered yet or, proxy-service is modified (md5 changed),
+        // add metadata files to temp folder
+        if (StringUtils.isEmpty(md5FromServer) || !newMD5String.equals(md5FromServer)) {
+            if (!newMetaFile.mkdir()) {
+                log.error("Could not create temporary files");
+                return false;
+            }
+            // Copy metadata yaml file to the temp location.
+            FileUtils.copyFile(metadataYamlFile, newYamlFile);
+            // Copy wsdl file to the temp location.
+            FileUtils.copyFile(wsdlFile, new File(newMetaFile, wsdlFile.getName()));
+
+            // Add to map to be included in filter formdata field.
+            Map<String, String> md5Map = new HashMap<>();
+            md5Map.put(METADATA_KEY, key);
+            // if API is changed add the previous MD5 sum, if new API add the new MD5 sum.
+            md5Map.put(MD5, StringUtils.isEmpty(md5FromServer) ? newMD5String : md5FromServer);
```

Comment:
Do we really need a list of maps here? Can't we simply use an object instead of the map and add it to the list? Or atleast limit the size of the map since we know that it is 2?

Reply:
I used the same logic as in APIs https://github.com/wso2/micro-integrator/blob/master/components/org.wso2.micro.integrator.initializer/src/main/java/org/wso2/micro/integrator/initializer/utils/ServiceCatalogUtils.java#L416 to maintain the same behaviour.